File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/FindKthLargest.java
package com.only4.algorithm.version4j.leetcode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Random;

public class FindKthLargest {
    public int findKthLargest(int[] nums, int k) {
        return quickSelect(Arrays.asList(Arrays.stream(nums).boxed().toArray(Integer[]::new)), k);
    }

    private int quickSelect(List<Integer> nums, int k) {
        Random random = new Random();
        int pivot = nums.get(random.nextInt(nums.size()));

        List<Integer> larger = new ArrayList<>();
        List<Integer> equal = new ArrayList<>();
        List<Integer> smaller = new ArrayList<>();

        for (int num : nums) {
            if (num > pivot) {
                larger.add(num);
            } else if (num == pivot) {
                equal.add(num);
            } else {
                smaller.add(num);
            }
        }

        if (k <= larger.size()) {
            return quickSelect(larger, k);
        } else if (k <= larger.size() + equal.size()) {
            return pivot;
        } else {
            return quickSelect(smaller, k - larger.size() - equal.size());
        }
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/findKthLargest.kt
package com.only4.algorithm.version4k.leetcode

import kotlin.random.Random

fun findKthLargest(nums: IntArray, k: Int): Int {
    fun quickSelect(nums: List<Int>, k: Int): Int {
        val pivot = nums[Random.nextInt(nums.size)]

        val larger = mutableListOf<Int>()
        val equal = mutableListOf<Int>()
        val smaller = mutableListOf<Int>()

        for (num in nums) {
            when {
                num > pivot -> larger.add(num)
                num == pivot -> equal.add(num)
                else -> smaller.add(num)
            }
        }

        return when {
            k <= larger.size -> quickSelect(larger, k)
            k <= larger.size + equal.size -> pivot
            else -> quickSelect(smaller, k - larger.size - equal.size)
        }
    }

    return quickSelect(nums.toList(), k)
}
